logstash_async/transport.py

```diff
@@ -176,3 +224,117 @@ def send(self, events, use_logging=False):
+    with certifi. If you pass a string with a file location to CA certificate
+    the class tries to validate it against it. (Default: True)
+    :type ssl_verify: bool or str
+    :param username: Username for basic authorization (Default: "")
+    :type username: str
+    :param password: Password for basic authorization (Default: "")
+    :type password: str
+    """
+
+    def __init__(
+            self,
+            host,
+            port,
+            timeout=2.0,
+            ssl_enable=True,
+            ssl_verify=True,
+            **kwargs
+    ):
+        super().__init__(host, port, timeout, ssl_enable, ssl_verify)
+        self.username = kwargs.get('username', None)
+        self.password = kwargs.get('password', None)
+        self.__session = None
+        self.__max_attempts = 3
+
+    @property
+    def url(self):
+        """The URL of the logstash pipeline based on the hostname, the port and
+        the TLS usage.
+
+        :return: The URL of the logstash pipeline
+        :rtype: str
+        """
+        protocol = 'http'
+        if self.ssl_enable:
+            protocol = 'https'
+        return '{}://{}:{}'.format(protocol, self.host, self.port)
+
+    def encode(self, events):
+        """Decodes a list of events
+        :param events: A list of events
+        :type events: list
+        :return: A list of decoded events
+        :rtype: str
+        """
+        return [json.loads(event) for event in events]
+
+    def __auth(self):
+        """The authentication method for the logstash pipeline. If the username
+        or the password is not set correctly it will return None.
+
+        :return: A HTTP basic auth object or None
+        :rtype: HTTPBasicAuth
+        """
+        if self.username is None or self.password is None:
+            return None
+        return HTTPBasicAuth(self.username, self.password)
+
+    def close(self):
+        """The HTTP connection does not need to be closed because it's
+        stateless.
+        """
+        if self.__session is not None:
+            self.__session.close()
+
+    def __backoff(self, attempt, cap=3000, base=10):
+        return random.randrange(0, min(cap, base * 2 ** attempt))
+
+    def send(self, events, **kwargs):
+        """Send events to the logstash pipeline
+
+        :param events: A list of events
+        :type events: list
+        :param use_logging: Not used!
+        :type use_logging: bool
+        """
+        headers = {'Content-Type': 'application/json'}
+        self.__session = requests.Session()
+        attempt = 0
+        while attempt < self.__max_attempts:
+            response = requests.post(
+                self.url,
+                headers=headers,
+                json=self.encode(events),
+                verify=self.ssl_verify,
+                auth=self.__auth())
+            status_code = response.status_code
+            if status_code == 200:
+                break
+            if status_code == 429:
```

Comment:
Basically, we don't need to handle 429 and backoff explicitly.
The caller (`LogProcessingWorker`) will by design handle errors which are raised here and requeue all events passed to `send()`. The other transports also don't have own error handling as it is done by the caller more gracefully.

Reply:
What exception should be raised in case of an error?

Reply:
It doesn't matter much which exception is thrown. See https://github.com/eht16/python-logstash-async/blob/master/logstash_async/worker.py#L212, `Exception` is catched.
Raising a `RuntimeError` as you did should be fine.

The idea of the exception handling in the worker (as mentioned above) is that the worker will detect if there was an error on delivery via the transport and in this case, the events are put back to the database so that delivery will be tried again. By putting the events back to the database, they are also persisted in case the application shuts down between the failed delivery attempt and the next attempt.
While there is no real backoff logic, the `QUEUED_EVENTS_FLUSH_` settings in `constants.py` can be used to tune the sending behavior.
And this works for all transports in the same way.

Reply:
It might be even easier to just use `response.raise_for_status()`, this will raise a proper exception for any non-success status. The exceptions look like:
```
requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://example.com/non-existant
requests.exceptions.HTTPError: 403 Client Error: Forbidden for url: https://example.com/admin/
```
So not message composition and status code handling is necessary.
Just an idea.

Reply:
I like the idea. :-)